rfc/rbac-ng.md

```diff
@@ -0,0 +1,154 @@
+Authentication is handled by Gate typically, so this will reuse the existing `X-Spinnaker-User` header for denoting the principal of the request.
+User details are tracked by Fiat, so this data will need to be mirrored into OPA as a new backend.
+Authorization is the more complex part as permission data is stored by the various microservices that own the resources.
+These checks are typically handled through the `PermissionEvaluator` interface from Spring Security which takes three arguments: the authenticated user, the resource, and the permission.
+Permission evaluation will be given the context of what roles the requesting user has, what permissions apply to applications, and what permissions apply to accounts.
+As applications are checked by multiple services, these permissions will be pushed by Front50 into Fiat so that any service can use application permissions.
+External role data can either be pushed from Gate on login (and when service accounts are saved) or included as part of the input data to Fiat for policy decisions depending on if it's locally available (like being included as JWT metadata or X.509 attributes).
+The Fiat API will be updated to support permission checks on these new resources along with supporting role policies.
+
+Account permissions are most applicable to metric store accounts where only the `READ` and `WRITE` permissions apply.
+Granting read permission to a metric store account allows a user to fetch metrics by querying its `MetricsService` bean.
+Granting write permission to a metric store account allows a sid to manage permissions of the account, update the account configuration, and delete the account.
+Note that write permissions in this context are unrelated to writing metrics to a metrics store; metrics should be written to stores from the deployed applications being measured.
+Access controls will be defined in terms of role policies which are intended to replace direct use of permissions maps.
+
+Credentials management will be implemented through a SQL provider for managing metric store account definitions along with integration with user secrets.
+When credentials definitions are parsed into account credentials, the parser can replace user secret URIs with their decrypted data if the calling user has permission to read them.
+These credentials can be combined with existing configuration file provided accounts.
+A REST API will be added for saving and deleting credentials with the former providing an API for managing the permissions of said credentials.
+
+### Dependencies
+
+Open Policy Agent would be a new dependency from outside the project.
+OPA is commonly used with a lot of the technology that Spinnaker integrates with.
+This is intended to replace most of what Fiat currently does by externalizing authorization decisions which can enable massive code simplifications to Fiat if fully embraced.
```

Comment:
Do we plan to bundle an OPA distribution as part of a Spinnaker install, for example using Halyard or Armory's operator? Or are we saying an installation of OPA will become a requirement for using Fiat?

Granted installing OPA on Kubernetes is trivial, but it's a extra hopp people would have to jump through.

Reply:
Yeah, I plan to update Halyard at least because this would likely be a sidecar for Fiat. I can also look at the other distribution mechanisms we have here since like you said, OPA is fairly straightforward to deploy.